examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+
+
+def create_tf_dataset(data, bs=4):
+    audio_ds = create_audio_ds(data)
+    text_ds = create_text_ds(data)
+    ds = tf.data.Dataset.zip((audio_ds, text_ds))
+    ds = ds.map(lambda x, y: {"src": x, "trg": y})
+    ds = ds.batch(bs)
+    ds = ds.prefetch(tf.data.experimental.AUTOTUNE)
+    return ds
+
+
+"""
+Check contents of one batch
+"""
+
+
+split = int(len(data) * 0.99)
+train_data = data[:split]
+test_data = data[split:]
+
+ds = create_tf_dataset(train_data, bs=64)
+val_ds = create_tf_dataset(test_data, bs=4)
+
+for i in ds.take(1):
+    print(i["src"].shape)
+    print(i["trg"])
+
+
+"""
+## Callbacks to display predictions and to change learning rate
+"""
+
+
+class DisplayOutputs(keras.callbacks.Callback):
+    def __init__(
+        self, batch, idx_to_token, trg_start_token_idx=27, trg_end_token_idx=28
+    ):
+        """ Displays a batch of outputs after every epoch 
+        Arguments:
+        - batch: test batch containing the keys "src" and "trg"
+        - idx_to_token: a List containing the vocabulary tokens corresponding to their indices
+        - trg_start_token_idx: start token index in the target vocabulary
+        - trg_end_token_idx: end token index in the target vocabulary
+        """
+        self.batch = batch
+        self.trg_start_token_idx = trg_start_token_idx
+        self.trg_end_token_idx = trg_end_token_idx
+        self.idx_to_char = idx_to_token
+
+    def on_epoch_end(self, epoch, logs=None):
+        if epoch % 5 != 0:
+            return
+        src = self.batch["src"]
+        trg = self.batch["trg"].numpy()
+        bs = tf.shape(src)[0]
+        preds = self.model.generate(src, self.trg_start_token_idx)
+        preds = preds.numpy()
+        for i in range(bs):
+            target = ""
+            for idx in trg[i, :]:
+                target += "" + self.idx_to_char[idx]
+            prediction = ""
+            over = False
+            for idx in preds[i, :]:
+                if over:  # Add padding token once end token has beeen predicted
+                    prediction += "-"
+                    continue
+                if idx == self.trg_end_token_idx:
+                    over = True
+                prediction += "" + self.idx_to_char[idx]
+            print(f"target:     {target}")
+            print(f"prediction: {prediction}")
+            print()
+
+
+def scheduler(epoch, lr):
+    """ linear warm up - linear decay """
+    init_lr = 0.00001
+    lr_after_warmup = 0.001
+    final_lr = 0.00001
+    warmup_epochs = 15
+    decay_epochs = 85
+    if epoch < warmup_epochs:
+        return init_lr + ((lr_after_warmup - init_lr) / (warmup_epochs - 1)) * epoch
+    return max(
+        final_lr,
+        lr_after_warmup
+        - (epoch - warmup_epochs) * (lr_after_warmup - final_lr) / (decay_epochs),
+    )
+
+
+"""
+## Create & train the end-to-end model
+"""
+
+
+for i in val_ds.take(1):
+    batch = i  # Use the first batch of validation set to display outputs
```

Comment:
The purpose of these 2 lines is unclear, please clarify

Reply:
These are just to fetch a single batch from the validation set for the display callback.
It can be replaced by `batch = next(iter(val_ds))[1]`, if that feels better.